Mathlib/Algebra/Hom/Iterate.lean

```diff
@@ -0,0 +1,267 @@
+
+theorem iterate_map_smul (n m : ℕ) (x : R) : (f^[n]) (m • x) = m • (f^[n]) x :=
+  f.toAddMonoidHom.iterate_map_smul n m x
+#align ring_hom.iterate_map_smul RingHom.iterate_map_smul
+
+end Semiring
+
+variable {R : Type _} [Ring R] (f : R →+* R) (n : ℕ) (x y : R)
+
+theorem iterate_map_sub : (f^[n]) (x - y) = (f^[n]) x - (f^[n]) y :=
+  f.toAddMonoidHom.iterate_map_sub n x y
+#align ring_hom.iterate_map_sub RingHom.iterate_map_sub
+
+theorem iterate_map_neg : (f^[n]) (-x) = -(f^[n]) x :=
+  f.toAddMonoidHom.iterate_map_neg n x
+#align ring_hom.iterate_map_neg RingHom.iterate_map_neg
+
+theorem iterate_map_zsmul (n : ℕ) (m : ℤ) (x : R) : (f^[n]) (m • x) = m • (f^[n]) x :=
+  f.toAddMonoidHom.iterate_map_zsmul n m x
+#align ring_hom.iterate_map_zsmul RingHom.iterate_map_zsmul
+
+end RingHom
+
+theorem Equiv.Perm.coe_pow {α : Type _} (f : Equiv.Perm α) (n : ℕ) : ⇑(f ^ n) = f^[n] :=
+  hom_coe_pow _ rfl (fun _ _ => rfl) _ _
+#align equiv.perm.coe_pow Equiv.Perm.coe_pow
+
+--what should be the namespace for this section?
+section Monoid
+
+variable [Monoid G] (a : G) (n : ℕ)
+
+@[simp, to_additive]
+theorem smul_iterate [MulAction G H] : (a • · : H → H)^[n] = (a ^ n • ·) :=
+  funext fun b =>
+    Nat.recOn n (by rw [iterate_zero, id.def, pow_zero, one_smul])
+    fun n ih => by rw [iterate_succ', comp_apply, ih, pow_succ, mul_smul]
+#align smul_iterate smul_iterate
+#align vadd_iterate vadd_iterate
+
+@[simp, to_additive]
+theorem mul_left_iterate : (a * ·)^[n] = (a ^ n * ·) :=
+  smul_iterate a n
+#align mul_left_iterate mul_left_iterate
+#align add_left_iterate add_left_iterate
+
+@[simp, to_additive]
+theorem mul_right_iterate : (· * a)^[n] = (· * a ^ n) :=
+  smul_iterate (MulOpposite.op a) n
+#align mul_right_iterate mul_right_iterate
+#align add_right_iterate add_right_iterate
+
+@[to_additive]
+theorem mul_right_iterate_apply_one : ((· * a)^[n]) 1 = a ^ n := by simp [mul_right_iterate]
+#align mul_right_iterate_apply_one mul_right_iterate_apply_one
+#align add_right_iterate_apply_zero add_right_iterate_apply_zero
+
+@[simp, to_additive]
+theorem pow_iterate (n : ℕ) (j : ℕ) : (fun x : G => x ^ n)^[j] = fun x : G => x ^ n ^ j :=
+  letI : MulAction ℕ G :=
+    { smul := fun n g => g ^ n
+      one_smul := pow_one
+      mul_smul := fun m n g => pow_mul' g m n }
+  smul_iterate n j
+#align pow_iterate pow_iterate
+
+end Monoid
+
+section Group
+
+variable [Group G]
+
+@[simp, to_additive]
+theorem zpow_iterate (n : ℤ) (j : ℕ) : (fun x : G => x ^ n)^[j] = fun x => x ^ n ^ j :=
+  letI : MulAction ℤ G :=
+    { smul := fun n g => g ^ n
+      one_smul := zpow_one
+      mul_smul := fun m n g => zpow_mul' g m n }
+  smul_iterate n j
+#align zpow_iterate zpow_iterate
+#align zsmul_iterate zsmul_iterate
+
+end Group
+
+section Semigroup
+
+variable [Semigroup G] {a b c : G}
+
+-- Porting note: need `dsimp only`, see https://leanprover.zulipchat.com/#narrow/stream/
+-- 287929-mathlib4/topic/dsimp.20before.20rw/near/317063489
```

Comment:
Yeah, the issue is https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/dsimp.20before.20rw/near/317064195 : `(a * ·) t` expands to `(fun x ↦ a * x) t` in lean 4, but `((*)a) t` expands to `a * t` in lean 3, meaning that sometimes you have to `dsimp only` in Lean 4 to get back on track. You could mention this in the porting note but that's what's going on.